// effector-factorio/dist/esm/index.js
import { createContext, useContext } from 'react';
import { is } from 'effector';
import { jsx } from 'react/jsx-runtime';

function removeNonUnit(shape) {
  return Object.keys(shape).reduce((acc, key) => {
    const value = shape[key];
    if (is.unit(value)) {
      // @ts-expect-error
      acc[key] = value;
    }
    return acc;
  }, {});
}

function removeStorePrefix(shape) {
  return Object.keys(shape).reduce((acc, key) => {
    const unprefixedKey = key.replace(/^\$/, '');
    // @ts-expect-error
    acc[unprefixedKey] = shape[key];
    return acc;
  }, {});
}

const modelFactory = creator => {
  const ModelContext = /*#__PURE__*/createContext(null);
  const useModel = () => {
    const model = useContext(ModelContext);
    if (!model) {
      throw new Error('No model found');
    }
    return model;
  };
  const unitShape = () => {
    const model = useModel();
    const modelUnits = removeNonUnit(model);
    return removeStorePrefix(modelUnits);
  };
  return {
    /** Function that returns new `model` instance */
    createModel: creator,
    /** Hook that returns current `model` instance */
    useModel,
    /** `Provider` to pass current `model` instance into */
    Provider: ModelContext.Provider,
    '@@unitShape': unitShape
  };
};
/**
 * HOC that wraps your `View` into model `Provider`. Also adds `model` prop that will be passed into `Provider`
 * @param factory Factory that will be passed through Context
 * @param View Root component that will be wrapped into Context
 * @returns Wrapped component
 */
const modelView = (factory, View) => {
  const Render = props => {
    return /*#__PURE__*/jsx(factory.Provider, {
      value: props.model,
      children: /*#__PURE__*/jsx(View, {
        ...props
      })
    });
  };
  // `as` is used for a better "Go To Definition"
  return Render;
};

export { modelFactory, modelView };
