// effector-factorio/dist/cjs/solid.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var web = require('solid-js/web');
var solidJs = require('solid-js');
var effector = require('effector');

function removeNonUnit(shape) {
  return Object.keys(shape).reduce((acc, key) => {
    const value = shape[key];
    if (effector.is.unit(value)) {
      // @ts-expect-error
      acc[key] = value;
    }
    return acc;
  }, {});
}

function removeStorePrefix(shape) {
  return Object.keys(shape).reduce((acc, key) => {
    const unprefixedKey = key.replace(/^\$/, '');
    // @ts-expect-error
    acc[unprefixedKey] = shape[key];
    return acc;
  }, {});
}

/**
 * Creates model `factory`
 * @param creator Function that returns new `model` instance
 * @returns Factory with `createModel` method, `useModel` hook and model `Provider`
 */
const modelFactory = creator => {
  const ModelContext = solidJs.createContext(null);
  const useModel = () => {
    const model = solidJs.useContext(ModelContext);
    if (!model) {
      throw new Error('No model found');
    }
    return model;
  };
  const unitShape = () => {
    const model = useModel();
    const modelUnits = removeNonUnit(model);
    return removeStorePrefix(modelUnits);
  };
  return {
    /** Function that returns new `model` instance */
    createModel: creator,
    /** Hook that returns current `model` instance */
    useModel,
    /** `Provider` to pass current `model` instance into */
    Provider: ModelContext.Provider,
    '@@unitShape': unitShape
  };
};
/**
 * HOC that wraps your `View` into model `Provider`. Also adds `model` prop that will be passed into `Provider`
 * @param factory Factory that will be passed through Context
 * @param View Root component that will be wrapped into Context
 * @returns Wrapped component
 */
const modelView = (factory, View) => {
  const Render = props => {
    return web.createComponent(factory.Provider, {
      get value() {
        return props.model;
      },
      get children() {
        return web.createComponent(web.Dynamic, web.mergeProps({
          component: View
        }, props));
      }
    });
  };
  // `as` is used for a better "Go To Definition"
  return Render;
};

exports.modelFactory = modelFactory;
exports.modelView = modelView;
